### ISSUE TITLE
i/throw-illegal produces invalid code for CLJS

### ISSUE THREAD

**Issue by thheller** [nathanmarz/specter#267]
The `com.rpl.specter.impl/throw-illegal` macro produces invalid code for CLJS on the JVM (none self-hosted).

The generated code ends up with a reference to `java.lang.IllegalArgumentException`.

```
com.rpl.specter.NAME_transform_STAR_ = (function com$rpl$specter$NAME_transform_STAR_(structure,next_fn){
var new_name = (function (){var G__36732 = cljs.core.name(structure);
return (next_fn.cljs$core$IFn$_invoke$arity$1 ? next_fn.cljs$core$IFn$_invoke$arity$1(G__36732) : next_fn.call(null,G__36732));
})();
var ns = cljs.core.namespace(structure);
if((structure instanceof cljs.core.Keyword)){
return cljs.core.keyword.cljs$core$IFn$_invoke$arity$2(ns,new_name);
} else {
if((structure instanceof cljs.core.Symbol)){
return cljs.core.symbol.cljs$core$IFn$_invoke$arity$2(ns,new_name);
} else {
throw (new java.lang.IllegalArgumentException(com.rpl.specter.impl.smart_str.cljs$core$IFn$_invoke$arity$variadic(cljs.core.prim_seq.cljs$core$IFn$_invoke$arity$2(["NAME can only be used on symbols or keywords - ",structure], 0))));

}
}
});
```

```
------ WARNING #2 --------------------------------------------------------------
 File: com/rpl/specter.cljc:1288:19
--------------------------------------------------------------------------------
1285 |           new-ns (next-fn (namespace structure))]
1286 |       (cond (keyword? structure) (keyword new-ns name)
1287 |             (symbol? structure) (symbol new-ns name)
1288 |             :else (i/throw-illegal "NAMESPACE can only be used on symbols or keywords - " structure)
-------------------------^------------------------------------------------------
 Use of undeclared Var com.rpl.specter/java
--------------------------------------------------------------------------------
1289 |             ))))
1290 | 
1291 | (defdynamicnav
1292 |   ^{:doc "Adds the result of running select with the given path on the
--------------------------------------------------------------------------------
```

The problem is that the macro must decide at runtime whether to emit code for CLJS or CLJ as the reader conditional is no help there.

```
#?(
   :clj
   (defmacro throw-illegal [& args]
     `(throw* IllegalArgumentException ~@args))


   :cljs
   (defn throw-illegal [& args]
     (throw (js/Error. (apply str args)))))
```

**Comment by thheller** [comment]
FWIW the problem is that the macro version shadows the fn version due to https://dev.clojure.org/jira/browse/CLJS-2454.

It goes by unnoticed in normal CLJS since every symbol with a `.` in it is blindly accepted which `shadow-cljs` now properly validates and warns about. See https://dev.clojure.org/jira/browse/CLJS-712

**Comment by loganpowell** [comment]
Also experiencing this issue

**Comment by vigilancetech-com** [comment]
same with me.   May be unrelated but my app used to run and now doesn't.   Here's the js console output:

```
Installing CLJS DevTools 0.9.10 and enabling features :formatters :hints :async
browser.cljs:25 shadow-cljs: WebSocket connected!
browser.cljs:25 shadow-cljs: REPL session start successful
util.cljs:136 CLJS DevTools: some custom formatters were not rendered.
https://github.com/binaryage/cljs-devtools/blob/master/docs/faq.md#why-some-custom-formatters-were-not-rendered
devtools$util$check_custom_formatters_active_BANG_ @ util.cljs:136
core.cljs:5907 Uncaught RangeError: Maximum call stack size exceeded
    at cljs$core$build_subvec (core.cljs:5907)
    at cljs.core.Subvec.cljs$core$IWithMeta$_with_meta$arity$2 (core.cljs:5793)
    at Object.cljs$core$_with_meta [as _with_meta] (core.cljs:686)
    at Object.cljs$core$with_meta [as with_meta] (core.cljs:2059)
    at Object.cljs$core$vec [as vec] (core.cljs:5662)
    at cljs.core.Subvec.com$rpl$specter$impl$CoercePath$coerce_path$arity$1 (impl.cljc:180)
    at Object.com$rpl$specter$impl$coerce_path [as coerce_path] (impl.cljc:155)
    at cljs.core.Subvec.com$rpl$specter$impl$CoercePath$coerce_path$arity$1 (impl.cljc:180)
    at Object.com$rpl$specter$impl$coerce_path [as coerce_path] (impl.cljc:155)
    at cljs.core.Subvec.com$rpl$specter$impl$CoercePath$coerce_path$arity$1 (impl.cljc:180)
```

**Comment by nathanmarz** [comment]
Fixed in master and released as 1.1.3-SNAPSHOT. Please try that out and let me know how it goes.

**Comment by pepe** [comment]
Works great for me. Thank you very much!

**Comment by loganpowell** [comment]
Can confirm `1.1.3-SNAPSHOT` compiles (to JavaScript) without warnings. 👍

**Comment by tslocke** [comment]
@nathanmarz any sign of a 1.1.3 release?

**Comment by nathanmarz** [comment]
@tslocke Just released 1.1.3
